FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt

```diff
@@ -0,0 +1,424 @@
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.bouncycastle.openpgp.PGPPublicKeyRingCollection
+import org.bouncycastle.openpgp.PGPSecretKeyRingCollection
+import org.pgpainless.util.Passphrase
+import java.io.ByteArrayOutputStream
+import java.io.FileNotFoundException
+import java.io.InputStream
+import java.net.SocketException
+import java.util.Base64
+import java.util.Properties
+import javax.mail.Message
+import javax.mail.MessagingException
+import javax.mail.Multipart
+import javax.mail.Session
+import javax.mail.internet.InternetAddress
+import javax.mail.internet.MimeBodyPart
+import javax.mail.internet.MimeMessage
+import javax.net.ssl.SSLException
+import kotlin.random.Random
+
+/**
+ * @author Denis Bondarenko
+ *         Date: 12/29/21
+ *         Time: 9:30 AM
+ *         E-mail: DenBond7@gmail.com
+ */
+class HandlePasswordProtectedMsgWorker(context: Context, params: WorkerParameters) :
+  BaseMsgWorker(context, params) {
+  override suspend fun doWork(): Result = withContext(Dispatchers.IO) {
+    LogsUtil.d(TAG, "doWork")
+
+    if (isStopped) {
+      return@withContext Result.success()
+    }
+
+    try {
+      val account = AccountViewModel.getAccountEntityWithDecryptedInfoSuspend(
+        roomDatabase.accountDao().getActiveAccountSuspend()
+      ) ?: return@withContext Result.success()
+
+      val passwordProtectedCandidates = roomDatabase.msgDao().getOutboxMsgsByStatesSuspend(
+        account = account.email,
+        msgStates = listOf(MessageState.NEW_PASSWORD_PROTECTED.value)
+      )
+
+      if (passwordProtectedCandidates.isNotEmpty()) {
+        prepareAndUploadPasswordProtectedMsgsToFES(account)
+      }
+
+      return@withContext rescheduleIfActiveAccountWasChanged(account)
+    } catch (e: Exception) {
+      e.printStackTrace()
+      return@withContext Result.failure()
+    } finally {
+      LogsUtil.d(TAG, "work was finished")
+    }
+  }
+
+  private suspend fun prepareAndUploadPasswordProtectedMsgsToFES(account: AccountEntity) =
+    withContext(Dispatchers.IO) {
+      val apiRepository = FlowcryptApiRepository()
+      val keysStorage = KeysStorageImpl.getInstance(applicationContext)
+      val accountSecretKeys = PGPSecretKeyRingCollection(keysStorage.getPGPSecretKeyRings())
+
+      var list: List<MessageEntity>
+      var lastMsgUID = 0L
+      while (true) {
+        list = roomDatabase.msgDao().getOutboxMsgsByStatesSuspend(
+          account = account.email,
+          msgStates = listOf(MessageState.NEW_PASSWORD_PROTECTED.value)
+        )
+
+        if (list.isEmpty()) break
+        val msgEntity = list.firstOrNull { it.uid > lastMsgUID } ?: list[0]
+
+        lastMsgUID = msgEntity.uid
+
+        try {
+          if (msgEntity.isEncrypted == true && msgEntity.isPasswordProtected) {
+            //get msg attachments(including forwarded that can be encrypted)
+            val attachments = getAttachments(msgEntity)
```

Comment:
Thank you for the comment - this helps a lot.

It seems to me that the encrypted forwarded attachments are put inside of the `plainMimeMsgWithAttachments` without decrypting them first. I think that will make forwarded attachments not readable for the recipient on FES. Could you test?

If it does work correctly, why is it so? :-)

Reply:
> If it does work correctly, why is it so? :-)

it's a magic. I'm joking. Please look at 

`getMimeMessage -> EmailUtil.createMimeMsg() -> genBodyPartWithAtt() -> ForwardedAttachmentInfoDataSource().`

We decrypt all attachments before sending. In that place all attachments are encrypted, because in `PrepareOutgoingMessagesJobIntentService` we create an encrypted message(and encrypt all attachments). Or(And) after `ForwardedAttachmentsDownloaderWorker` we have an encrypted forwarded attachment.

> Could you test?

I've tested. Works well

Reply:
I think you've described it for the pgp-recipient. But for recipient who is reading it on the web portal, it must be "plain mime message" -> encrypt whole -> upload. But here it looks like it will be "plain mime message with encrypted attachments" -> encrypt whole -> upload. Which wouldn't work.

Even if it works, it really is confusing, especially for the web portal attachments scenario. We'll have to refactor it in the future to find a a more obvious structure.

Reply:
> But here it looks like it will be "plain mime message with encrypted attachments" -> encrypt whole -> upload. Which wouldn't work.

No, it's not.

> "plain mime message" -> encrypt whole -> upload

It's true.

I understand your questions and your opinion. As I said earlier all things here look a little difficult. Because we are talking about not simple things that were integrated in the existing logic(encrypted things where we are expection plain data) + Android things = looks difficult.

I will add more details a bit later. It will take some time

Reply:
So let's start :)

First of all, let me remind you how the current process of sending a message works

1. a user creates an outgoing message and clicks on the `Send` button
2. the app creates [OutgoingMessageInfo](https://github.com/FlowCrypt/flowcrypt-android/blob/9ff16025c328de0f0e873f81ce342db02542a8a0/FlowCrypt/src/main/java/com/flowcrypt/email/api/email/model/OutgoingMessageInfo.kt#L23). This object contains outgoing message info(like `to, cc, bcc, subject, message text` and so on). Then `CreateMessageFragment` enqueues a new task for [PrepareOutgoingMessagesJobIntentService](https://github.com/FlowCrypt/flowcrypt-android/blob/bb1a66b01e049830ec4fa8b0cc75ce209eccdc4a/FlowCrypt/src/main/java/com/flowcrypt/email/service/PrepareOutgoingMessagesJobIntentService.kt)
3. `PrepareOutgoingMessagesJobIntentService` creates MIME message: encrypt text if needed, encrypt non-forwarded attachments if needed and store them in a message temporary folder. Then we add a MIME message without attachments to the local database(create a message in `Outbox`) + we store attachments (with [Uri](https://developer.android.com/reference/android/net/Uri), later it will be used to get access to attachment during the sending process) in the `attachments` table. Then if we have no forwarded attachments we mark such a message with [MessageState.QUEUED](https://github.com/FlowCrypt/flowcrypt-android/blob/c9797018be3f96a59bccb84767d437a0503a1198/FlowCrypt/src/main/java/com/flowcrypt/email/database/MessageState.kt#L24) or [MessageState.NEW_FORWARDED](https://github.com/FlowCrypt/flowcrypt-android/blob/c9797018be3f96a59bccb84767d437a0503a1198/FlowCrypt/src/main/java/com/flowcrypt/email/database/MessageState.kt#L29) if we have at least one forwarded attachment. if the message was marked as `MessageState.QUEUED` (means the message is ready to be sent) we run [MessagesSenderWorker](https://github.com/FlowCrypt/flowcrypt-android/blob/c72f6343ba3886376d8f9e2be311493ff9695208/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/MessagesSenderWorker.kt#L72) that sends outgoing messages. if the message was marked as `MessageState.NEW_FORWARDED` (means the message has forwarded type and has at least one forwarded attachment that is not downloaded) we run [ForwardedAttachmentsDownloaderWorker](https://github.com/FlowCrypt/flowcrypt-android/blob/bb1a66b01e049830ec4fa8b0cc75ce209eccdc4a/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/ForwardedAttachmentsDownloaderWorker.kt#L54) that downloads forwarded attachments.
4. `ForwardedAttachmentsDownloaderWorker` downloads forwarded attachments, save them to the message temporary folder and mark the message as `MessageState.QUEUED`. It means we already have all needed content and can send a message. Then we run `MessagesSenderWorker`
5. `MessagesSenderWorker` sends outgoing messages

It's a logic without nuances. Just to show how it works.

Now let's me show how it works after changes from the current PR.

1. a user creates an outgoing PASSWORD-PROTECTED message and clicks on the `Send` button
2. the app creates [OutgoingMessageInfo](https://github.com/FlowCrypt/flowcrypt-android/blob/9ff16025c328de0f0e873f81ce342db02542a8a0/FlowCrypt/src/main/java/com/flowcrypt/email/api/email/model/OutgoingMessageInfo.kt#L23). Then `CreateMessageFragment` enqueues a new task for [PrepareOutgoingMessagesJobIntentService](https://github.com/FlowCrypt/flowcrypt-android/blob/bb1a66b01e049830ec4fa8b0cc75ce209eccdc4a/FlowCrypt/src/main/java/com/flowcrypt/email/service/PrepareOutgoingMessagesJobIntentService.kt)
3. `PrepareOutgoingMessagesJobIntentService` creates MIME message: encrypt text as inlined attachment (`message.asc`) 
https://github.com/FlowCrypt/flowcrypt-android/blob/a7b0e7e02ee2e624802b316904a1f89030d655a4/FlowCrypt/src/main/java/com/flowcrypt/email/api/email/EmailUtil.kt#L1155

 , encrypt non-forwarded attachments if needed and store them in a message temporary folder. 

at this step MIME message without attachments will look like 

```
Date: Wed, 5 Jan 2022 21:25:17 +0200 (GMT+02:00)
From: den@flowcrypt.com
To: unknown@flowcrypt.com
Message-ID: <235793934.1.1641410718042@flowcrypt.com>
Subject: Test password-protected
MIME-Version: 1.0
Content-Type: multipart/mixed; 
	boundary="----=_Part_0_159234674.1641410708833"

------=_Part_0_159234674.1641410708833
Content-Type: application/pgp-encrypted; name=message.asc
Content-Transfer-Encoding: 7bit
Content-Description: OpenPGP encrypted message
Content-Disposition: inline; filename=message.asc

-----BEGIN PGP MESSAGE-----
Version: PGPainless

hF4DE4nP38Il0h4SAQdAn+LpwTvhJFWingurhm48emXoO8samS4O/46/CIhMY2gw
9i2oaiLBm2ZLpPGh4O0C/GHm340Mi+LGj/pfXgBSYmyP6ZLSpRj+R1snqW1HwRca
0rwBa480Pc3w4IyzfhiW4wnnLeUTkxxulg6rt1wB++VDGjWLvq/u2U7JnzF+E17m
h7oZto3cq5Mm+z0RwkdQxwvIfgjsa6C7go9667U56eL1cNH5CM4CpC/xwcZnrr51
MC4/g9c4TeDRprwTPgXCtvwKuQnX6IOCnB3EtbpmELqrjXKPyPuNo0kqbJlpmepm
Q5WnA7YJanYzpQ9hW7woMinQH6Q0T0W2kXYF9ZUC0NRoe8VmKtErfdD14v6FcQ==
=Fdto
-----END PGP MESSAGE-----

------=_Part_0_159234674.1641410708833--
```

Then we add this MIME message without attachments to the local database(create a message in `Outbox`) + we store attachments (with [Uri](https://developer.android.com/reference/android/net/Uri), later it will be used to get access to attachment during the sending process) in the `attachments` table. Then if we have no forwarded attachments we mark such a message with [MessageState.NEW_PASSWORD_PROTECTED](https://github.com/FlowCrypt/flowcrypt-android/blob/c9797018be3f96a59bccb84767d437a0503a1198/FlowCrypt/src/main/java/com/flowcrypt/email/database/MessageState.kt#L45) or [MessageState.NEW_FORWARDED](https://github.com/FlowCrypt/flowcrypt-android/blob/c9797018be3f96a59bccb84767d437a0503a1198/FlowCrypt/src/main/java/com/flowcrypt/email/database/MessageState.kt#L29) if we have at least one forwarded attachment. if the message was marked as `MessageState.NEW_PASSWORD_PROTECTED` we run [HandlePasswordProtectedMsgWorker](https://github.com/FlowCrypt/flowcrypt-android/blob/115ce5fe0c190c7b85d2d5ff605628943224c5dc/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt#L68). if the message was marked as `MessageState.NEW_FORWARDED` (means the message has forwarded type and has at least one forwarded attachment that is not downloaded) we run [ForwardedAttachmentsDownloaderWorker](https://github.com/FlowCrypt/flowcrypt-android/blob/bb1a66b01e049830ec4fa8b0cc75ce209eccdc4a/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/ForwardedAttachmentsDownloaderWorker.kt#L54) that downloads forwarded attachments. 

During adding a new outgoing message to the local database we also add password(encrpyted with AndroidKeyStore). Later, based on this info we can check if a message is password-protected.
4. `ForwardedAttachmentsDownloaderWorker` downloads forwarded attachments, save them to the message temporary folder and mark the message as `MessageState.NEW_PASSWORD_PROTECTED `. It means we already have all needed content and can process a password-protected message. Then we run `HandlePasswordProtectedMsgWorker`
5. `HandlePasswordProtectedMsgWorker` creates and sends a password-protected message to FES, and modifies existing MIME message that saved in the local database (adds URL). Let's me remind you that in the local database we have MIME message with ENCRYPTED content that was created in 3. Also we have encrypted attachments.

Then we construct `MimeMessage` using existing raw MIME message from the local database(message entity: our `Date: Wed, 5 Jan 2022 21:25:17 +0200 (GMT+02:00)...`) + encrypted attachments

![image](https://user-images.githubusercontent.com/2863246/148279287-7ea23178-48c2-4380-88b3-97d9edf42c13.png)

You've asked me a few times about this. Why `getDecryptedContentFromMessage`?

```
val originalText = getDecryptedContentFromMessage(
              mimeMsgWithAttachments = plainMimeMsgWithAttachments,
              accountSecretKeys = accountSecretKeys,
              keysStorage = keysStorage
            )
```

This code gets `message.asc` from our `plainMimeMsgWithAttachments` and decrypts it. Then we receive original text that was sent by a user on the message compose screen. we decrypt the following here

```
------=_Part_0_159234674.1641410708833
Content-Type: application/pgp-encrypted; name=message.asc
Content-Transfer-Encoding: 7bit
Content-Description: OpenPGP encrypted message
Content-Disposition: inline; filename=message.asc

-----BEGIN PGP MESSAGE-----
Version: PGPainless

hF4DE4nP38Il0h4SAQdAn+LpwTvhJFWingurhm48emXoO8samS4O/46/CIhMY2gw
9i2oaiLBm2ZLpPGh4O0C/GHm340Mi+LGj/pfXgBSYmyP6ZLSpRj+R1snqW1HwRca
0rwBa480Pc3w4IyzfhiW4wnnLeUTkxxulg6rt1wB++VDGjWLvq/u2U7JnzF+E17m
h7oZto3cq5Mm+z0RwkdQxwvIfgjsa6C7go9667U56eL1cNH5CM4CpC/xwcZnrr51
MC4/g9c4TeDRprwTPgXCtvwKuQnX6IOCnB3EtbpmELqrjXKPyPuNo0kqbJlpmepm
Q5WnA7YJanYzpQ9hW7woMinQH6Q0T0W2kXYF9ZUC0NRoe8VmKtErfdD14v6FcQ==
=Fdto
-----END PGP MESSAGE-----

------=_Part_0_159234674.1641410708833--
```
Move on

![image](https://user-images.githubusercontent.com/2863246/148280624-2e6448fc-115a-4fad-9b1d-7791d2e84f00.png)

At this step we should create raw MIME message that will be encrypted with our password. We take our `plainMimeMsgWithAttachments` that is a plain text that contains encrypted original data as `message.asc`. Before creating a raw MIME message we should modify `plainMimeMsgWithAttachments` a little. 

Firstl of all we should repelace a part that contains `message.asc` with `bodyWithReplyToken`. We do it here

https://github.com/FlowCrypt/flowcrypt-android/blob/115ce5fe0c190c7b85d2d5ff605628943224c5dc/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt#L254

After that we can construct `rawMimeMsg`. After that we do all rest things like encrypting and uploading, updating cached MIME message with URL from FES.

I need to clarify one more moment

> But for recipient who is reading it on the web portal, it must be "plain mime message" -> encrypt whole -> upload. But here it looks like it will be "plain mime message with encrypted attachments" -> encrypt whole -> upload. Which wouldn't work.

`EmailUtil.createMimeMsg()` is responsible for that. It knows that we should decrypt encrypted attachments before making `rawMimeMsg`. It does it on the fly. Without knowing a full logic it is a little hard to understand it.

https://github.com/FlowCrypt/flowcrypt-android/blob/a7b0e7e02ee2e624802b316904a1f89030d655a4/FlowCrypt/src/main/java/com/flowcrypt/email/api/email/EmailUtil.kt#L1020

This part does decryption

```
val attBodyPart = genBodyPartWithAtt(
            context = context,
            att = att,
            shouldBeEncrypted = msgEntity.isEncrypted ?: false,
            publicKeys = publicKeys,
            secretKeys = secretKeys,
            ringProtector = ringProtector
          )
```

After all of that we mark the message as `MessageState.QUEUED`. It means we already have all needed content and can send a message. Then we run `MessagesSenderWorker`

6. `MessagesSenderWorker` sends outgoing messages

P.S. I hope it will help to understand how it works.

Reply:
@DenBond7 I think the above should be put into some separate markdown file and checked in into the repo.

Reply:
Agree